RFC: sets/domains

tbreloff opened the pull request:
What do you think? Will this work for our purposes?

tbreloff committed:
sets/domains

tbreloff commented:
We might consider having an `outputdomain` in addition, for example with SVM isn't the output domain an `IntervalSet(-1,1)` but the target domain is `DiscreteSet([1,-1])`?  Though I guess the more correct way to do this is that the margin loss has `inputdomain(loss) == TupleSet(IntervalSet(-1,1), DiscreteSet([1,-1]))`.

Anyways... criticism welcome

Evizero commented:
> for example with SVM isn't the output domain an IntervalSet(-1,1)

No. The model is just a linear model like in linear regression. the interpretation is just different. So the output domain is R

EDIT:
The points that are not filled in in this plot will have an output greater than 1 or smaller than -1. all points between the dotted lines have an output in [-1,1]

![svm](http://docs.opencv.org/2.4/_images/optimal-hyperplane.png)

Evizero commented on src/LearnBase.jl:
```diff
@@ -117,6 +117,60 @@ function update! end
+end
+
+Base.rand{T<:Number}(s::IntervalSet{T}) = rand() * (s.hi - s.lo) + s.lo
+Base.in{T<:Number}(x::Number, s::IntervalSet{T}) = s.lo <= x <= s.hi
```
can you please add some tests for `in` and `rand`

Evizero commented:
I'll be greedy and ask for some tests. Other than that I like it 👍

tbreloff commented:
Yeah was planning on tests.  I just wanted opinions before I went to the trouble.

tbreloff committed:
sets: fixes, cleanup, multidimensional rand; added tests

tbreloff commented:
Cleaned up and tests added.  I'd be happy to merge if you are.  cc: @Evizero @ahwillia @spencerlyon2

Evizero commented:
Looks good to me in terms of content.

But before we merge, please change the tabs to 4 spaces in the tests :)

sglyon commented on src/LearnBase.jl:
```diff
@@ -117,6 +117,72 @@ function update! end
+TupleSet(sets::AbstractSet...) = TupleSet(sets)
+Base.rand(sets::TupleSet) = [rand(s) for s in sets.sets]
+function Base.rand(sets::TupleSet, dim1::Integer, dims::Integer...)
+    A = Array(typeof(rand(sets)), dim1, dims...)
```
Probably super minor, but you could save the first `rand(sets)` if you wanted

sglyon commented:
Sorry I didn't pitch in earlier -- I didn't have much to add. 

I would vote that we add a method for generating one sample from a `TupleSet` and put the output in a `Tuple` similar to what we did [here](https://github.com/tbreloff/Reinforce.jl/blob/da0dbfccfe8a6f4ee5427d628a1f67047a8884c4/src/Reinforce.jl#L84)

Otherwise, looks good to me!

tbreloff commented:
> you could save the first rand(sets) if you wanted

 Thought about this, though it makes the code uglier.  I think a better solution is to have a query method to static-ly return the type that rand will output.  Something like:

``` julia
randtype(s::IntervalSet) = Float64
```

and then use this to compute the randtype from the randtype of its components.

> I would vote that we add a method for generating one sample from a TupleSet and put the output in a Tuple similar to what we did here

Ok

Also... tests are failing on 0.4 because of types returned from array comprehensions.  What do we want to do?

```
ERROR: LoadError: test failed: Array{Union{Float64,Int64},1} == Array{Real,1}
```

tbreloff commented:
> What do we want to do?

One option is that we don't support 0.4.  Just one more motivating factor to get everyone on 0.5 as soon as possible.

tbreloff committed:
allow Tuple returns from TupleSet rand; tabs to spaces

Evizero commented:
if travis succeeds go ahead and merge. I just enabled appveyor, so if that one fails it may be for other reasons than this PR

Evizero commented:
looks like appveyor just succeeded on master, so it should hopefully succeed here as well

tbreloff committed:
added randtype

sglyon commented:
Should `randtype` be `Base.eltype`?

tbreloff commented:
Needs more review.  I added a method `randtype` which, as it sounds, returns the type that will be returned by `rand`.  I also went ahead and changed the logic to use group promotion so that we will create concrete types even when combining different set types.  For example:

``` julia
s = [IntervalSet(0,1), DiscreteSet([0,1])]
@test typeof(rand(s)) == Vector{Float64}
@test typeof(rand(s, 2)) == Vector{Vector{Float64}}
```

> Should randtype be eltype?

No.  See for example `IntervalSet(0,1)`

Evizero commented:
lgtm

Evizero commented:
nice thanks. Since this is now more than just definitions I should probably add coveralls as well
